libnopegl/src/node_filters.c

```diff
@@ -201,6 +226,97 @@ static int filteralpha_init(struct ngl_node *node)
     return register_resource(&s->filter.resources, "alpha", o->alpha_node, &o->alpha, NGLI_TYPE_F32);
 }
 
+static int filtercolormap_init(struct ngl_node *node)
+{
+    int ret = filter_init(node);
+    if (ret < 0)
+        return ret;
+
+    struct filtercolormap_priv *s = node->priv_data;
+    struct filtercolormap_opts *o = node->opts;
+
+    if (o->nb_colorkeys < 2) {
+        LOG(ERROR, "a minimum of 2 color keys is required to make a color map");
+        return NGL_ERROR_INVALID_ARG;
+    }
+
+    s->filter.name = "colormap";
+    s->filter.helpers = NGLI_FILTER_HELPER_MISC_UTILS | NGLI_FILTER_HELPER_LINEAR2SRGB | NGLI_FILTER_HELPER_SRGB2LINEAR;
+
+    /* Craft the shader dynamically according to the number of color keys */
+    struct bstr *str = ngli_bstr_create();
+    if (!str) {
+        ret = NGL_ERROR_MEMORY;
+        goto end;
+    }
+
+    #define I "    "
+
+    /* Prototype and initial declarations */
+    ngli_bstr_print(str, "vec4 filter_colormap(vec4 color, vec2 coords");
```

Comment:
It doesn't really help with long lines afaict. I could somehow split each format line, but at the same time I believe the flow of the generated code is more readable with single lines (because of the unrolling). In the source itself I can do various split, here are 2 different attempts for each block:

```diff
diff --git a/libnopegl/src/node_filters.c b/libnopegl/src/node_filters.c
index e38d16f47..83abc7aa0 100644
--- a/libnopegl/src/node_filters.c
+++ b/libnopegl/src/node_filters.c
@@ -273,18 +273,26 @@ static int filtercolormap_init(struct ngl_node *node)
     /* Identify left-most and right-most knots */
     for (size_t i = 0; i < o->nb_colorkeys; i++) {
         /* The '=' are here to make sure we enter in the loop at least once */
-        ngli_bstr_printf(str, I "if (pos%zu <= t_prv) { t_prv = pos%zu; v_prv = vec4(color%zu, opacity%zu); }\n", i, i, i, i);
-        ngli_bstr_printf(str, I "if (pos%zu >= t_nxt) { t_nxt = pos%zu; v_nxt = vec4(color%zu, opacity%zu); }\n", i, i, i, i);
+        ngli_bstr_printf(str, I "if (pos%zu <= t_prv) { "
+                                    "t_prv = pos%zu; "
+                                     "v_prv = vec4(color%zu, opacity%zu); "
+                                 "}\n"
+                              I "if (pos%zu >= t_nxt) { "
+                                     "t_nxt = pos%zu; "
+                                     "v_nxt = vec4(color%zu, opacity%zu); "
+                                 "}\n", i, i, i, i, i, i, i, i);
     }

     /* Identify the two closest surrounding knots (if any) */
     for (size_t i = 0; i < o->nb_colorkeys; i++) {
         /* The '=' are here to make sure we honor the knot if we are exactly on it */
-        ngli_bstr_printf(str, I "if (pos%zu <= t && pos%zu > t_prv) { t_prv = pos%zu; v_prv = vec4(color%zu, opacity%zu); }\n", i, i, i, i, i);
-        ngli_bstr_printf(str, I "if (pos%zu >= t && pos%zu < t_nxt) { t_nxt = pos%zu; v_nxt = vec4(color%zu, opacity%zu); }\n", i, i, i, i, i);
+        ngli_bstr_printf(str, I "if (pos%zu <= t && pos%zu > t_prv) { "
+                                "t_prv = pos%zu; v_prv = vec4(color%zu, opacity%zu); }\n", i, i, i, i, i);
+        ngli_bstr_printf(str, I "if (pos%zu >= t && pos%zu < t_nxt) { "
+                                "t_nxt = pos%zu; v_nxt = vec4(color%zu, opacity%zu); }\n", i, i, i, i, i);
     }
```

Overall, I don't find this to be an improvement because you loose the easy to read vertical diff between lines

Reply:
Ok.